cmd/calyptia/create_aggregator.go

```diff
@@ -0,0 +1,300 @@
+		Use:   "aggregator",
+		Short: "Create a new aggregator",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			home, err := os.UserHomeDir()
+			if err != nil {
+				return err
+			}
+
+			kubeconfig := filepath.Join(home, ".kube", "config")
+			k8sConfig, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+			if err != nil {
+				return err
+			}
+
+			clientset, err := kubernetes.NewForConfig(k8sConfig)
+			if err != nil {
+				return err
+			}
+
+			ctx := context.Background()
+
+			if aggregatorName != "" {
+				exists, err := config.aggregatorExistsWithName(ctx, aggregatorName)
+				if err != nil {
+					return err
+				}
+
+				if exists {
+					return fmt.Errorf("aggregator name %q taken", aggregatorName)
+				}
+			} else {
+				var err error
+				aggregatorName, err = generateAggregatorName()
+				if err != nil {
+					return err
+				}
+			}
+
+			if err := config.ensureNamespace(ctx, clientset, namespace); err != nil {
+				return err
+			}
+
+			clusterRole, err := config.createClusterRole(ctx, clientset, namespace, aggregatorName)
+			if err != nil {
+				return err
+			}
+
+			fmt.Printf("cluster role: %q\n", clusterRole.Name)
+
+			serviceAccount, err := config.createServiceAccount(ctx, clientset, namespace, aggregatorName)
+			if err != nil {
+				return err
+			}
+
+			fmt.Printf("service account: %q\n", serviceAccount.Name)
+
+			binding, err := config.createClusterRoleBinding(ctx, clientset, namespace, aggregatorName)
+			if err != nil {
+				return err
+			}
+
+			fmt.Printf("cluster role binding: %q\n", binding.Name)
+
+			deploy, err := config.createDeployment(ctx, clientset, namespace, aggregatorName)
+			if err != nil {
+				return err
+			}
+
+			fmt.Printf("deployment: %q\n", deploy.Name)
+
+			return nil
+		},
+	}
+
+	fs := cmd.Flags()
+	fs.StringVar(&aggregatorName, "name", "", "Aggregator name (autogenerated if empty)")
+	fs.StringVar(&namespace, "namespace", apiv1.NamespaceDefault, "Namespace")
+
+	return cmd
+}
+
+func (config *config) ensureNamespace(ctx context.Context, clientset *kubernetes.Clientset, namespace string) error {
+	exists, err := config.namespaceExists(ctx, clientset, namespace)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+
+	_, err = config.createNamespace(ctx, clientset, namespace)
+	return err
+}
+
+func (config *config) namespaceExists(ctx context.Context, clientset *kubernetes.Clientset, namespace string) (bool, error) {
+	_, err := clientset.CoreV1().Namespaces().Get(ctx, namespace, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		return false, nil
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (config *config) createNamespace(ctx context.Context, clientset *kubernetes.Clientset, namespace string) (*apiv1.Namespace, error) {
+	return clientset.CoreV1().Namespaces().Create(ctx, &apiv1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   namespace,
+			Labels: map[string]string{
+				// TODO: decide what to do
```

Comment:
So... this is making sure the namespace given by the user exists (if not, it creates it) but should it has some labels?
I mean, the other resources use a mix of the project ID and the generated aggregator name. But for the case of a k8s namespace... The namespace could be reused, dunno.

Give your opinions @niedbalski @gabrielbussolo

Reply:
well i think yes, as i got user can use namespace for other purposes also, nothing prevents him to do it.

Reply:
a minor thing, the *name* parameter is mandatory -- not optional -- and validated against the API before anything gets created at the k8s side.